docs/RFCS/20220426_rbac_extensions_system_privileges.md

```diff
@@ -0,0 +1,137 @@
+```
+CREATE TYPE system.object_type AS ENUM('system', 'jobs', 'changefeed_sink')
+```
+
+Note that this will be the first example of a system enum.
+
+
+The system.privileges table will store privileges for non-descriptor backed objects.
+We will require object_type + id to uniquely identify them. The privileges will be represented as a bit mask. We can serialize and deserialize synthetic privilege descriptors using rows in this table to leverage our existing code that manipulates PrivilegeDescriptors for grants. We can also consider making the privileges and role_options columns have type `[]STRING`, the downside here is that it takes slightly more space (probably not a concern) and that we have to add some logic to serialize the string array into a bit array which is how we currently manipulate privileges. 
+
+### Alternate considerations
+The system level grants could be stored on a PrivilegeDescriptor on the system database, however this would require us to support mutations on the system database descriptor which we currently do not do and this allows us to not perform a round trip whenever we look for a system table.
+
+We could create descriptors to represent every object we want to grant on. This is not an attractive option at first glance as these descriptors will likely only be used to store privileges and nothing more.
+
+## Syntax for System Level Privileges
+The current proposal is to support `GRANT SYSTEM [privileges...] TO [users...]`.
+We cannot do `GRANT [privileges…] ON SYSTEM TO [users…]` due to the fact that tables can be called system, this would be ambiguous. Using two words, such as SYSTEM ADMIN (can’t think of a good phrase here) would work.
+
+Proposed new system level privileges:
+VIEWACTIVITY (this can be broken down into more privileges)
+VIEWACTIVITYREDACTED
+CANCELQUERY
+MODIFYCLUSTERSETTING
+VIEWCLSUTERSETTING
+VIEWSYSTEMTABLES (can be broken down further if necessary)
+
+
+The syntax is up in the air however, the grant would create or modify the entry in the system.privileges table.
+
+Example: 
+`GRANT SYSTEM MODIFYCLUSTERSETTING TO foo` 
+
+Results in the row
+(foo, 1, “system”, 16384, 0)
+
+The bitmask value for MODIFYCLUSTERSETTING is assumed to be 2 << 13 = 16384 in the above example.
+
+Followed by 
+
+`GRANT SYSTEM VIEWCLUSTERSETTING TO foo` 
+
+Results in the row
+(foo, 1, “system”, 49152, 0)
+
+The bitmask value for VIEWCLUSTERSETTING is assumed to be 2 << 14 = 32768 in the above example. 32768 + 16384 = 49152.
+
+## Support for grants on virtual tables
+One thing worth explicitly calling out is that users have asked for the ability to grant on virtual tables (crdb_internal, pg_catalog). This proposal allows us to support grants on virtual tables and unify the grant experience for tables. Virtual tables have OIDs starting at `MaxUint32 = 1<<32 - 1` counting backwards, an INT (64 bit) column for object id supports this case.
```

Comment:
are virtual table oids fixed? i don't think they are today and we have shifted them between versions.

in particular, note https://github.com/cockroachdb/cockroach/blob/c80047e8cf31070dc446253640ebf643a394ef09/pkg/sql/catalog/catconstants/constants.go#L91-L384 and how we keep it in order and in the past have interrupted this order. how are we going to ensure these don't change over time in future if we are relying on it?

Reply:
I think it would be a non-issue to fix the IDs moving forward right?

Reply:
no. but preferably in a protobuf or something that makes it clear that changing it is bad.